src/main/java/io/jenkins/plugins/junit/checks/JUnitChecksPublisher.java

```diff
@@ -1,142 +1,145 @@
-        }
-
-
-        return builder.toString();
-    }
-}
+package io.jenkins.plugins.junit.checks;
+
+import edu.hm.hafner.util.VisibleForTesting;
+import hudson.model.TaskListener;
+import hudson.tasks.junit.CaseResult;
+import hudson.tasks.junit.TestResultAction;
+import hudson.tasks.junit.TestResultSummary;
+import io.jenkins.plugins.checks.api.ChecksConclusion;
+import io.jenkins.plugins.checks.api.ChecksDetails;
+import io.jenkins.plugins.checks.api.ChecksOutput;
+import io.jenkins.plugins.checks.api.ChecksPublisher;
+import io.jenkins.plugins.checks.api.ChecksPublisherFactory;
+import io.jenkins.plugins.checks.api.ChecksStatus;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.plugins.displayurlapi.DisplayURLProvider;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+@Restricted(NoExternalUse.class)
+public class JUnitChecksPublisher {
+    public static final String SEPARATOR = ", ";
+
+    // cap to avoid hitting check API message limit
+    private static final int MAX_MSG_SIZE_TO_CHECKS_API = 65535;
+    private final TestResultAction action;
+    private final TestResultSummary summary;
+
+    public JUnitChecksPublisher(final TestResultAction action, final TestResultSummary summary) {
+        this.action = action;
+        this.summary = summary;
+    }
+
+    public void publishChecks(TaskListener listener) {
+        ChecksPublisher publisher = ChecksPublisherFactory.fromRun(action.run, listener);
+        publisher.publish(extractChecksDetails());
+    }
+
+    @VisibleForTesting
+    ChecksDetails extractChecksDetails() {
+        String testsURL = DisplayURLProvider.get().getTestsURL(action.run);
+        ChecksOutput output = new ChecksOutput.ChecksOutputBuilder()
+                .withTitle(extractChecksTitle())
+                .withSummary("<sub>Send us [feedback](https://github.com/jenkinsci/junit-plugin/issues)")
+                .withText(extractChecksText(testsURL))
+                .build();
+
+        return new ChecksDetails.ChecksDetailsBuilder()
+                .withName(action.getChecksName())
+                .withStatus(ChecksStatus.COMPLETED)
+                .withConclusion(summary.getFailCount() > 0 ? ChecksConclusion.FAILURE : ChecksConclusion.SUCCESS)
+                .withDetailsURL(testsURL)
+                .withOutput(output)
+                .build();
+    }
+
+    private String extractChecksText(String testsURL) {
+        StringBuilder builder = new StringBuilder();
+        if (summary.getFailCount() > 0) {
+            List<CaseResult> failedTests = action.getResult().getFailedTests();
+
+            for (CaseResult failedTest: failedTests) {
+                String testReport = mapFailedTestToTestReport(failedTest);
+                int messageSize = testReport.getBytes(StandardCharsets.UTF_8).length + builder.toString().getBytes(StandardCharsets.UTF_8).length;

```

Comment:
is there any reason you're calling get bytes and not just `length()`

Reply:
Ideally `length()` should work perfectly fine, was thinking if text contains other more Unicode values length may not be equal to byte size

Reply:
i think length should be fine here. we're also substracting 1000 so should help any edge cases.

Reply:
makes sense, changed to `length()` 👍